### PR TITLE
euler1d gui and plotting area user experience enhancement

### PR DESCRIPTION
**Pull request by j8xixo12**
The following changes are included in this PR:

1. Plotting area user experience enhancement
   - Grid layout : all plot sharing same x and y axis label
   - Single layout : add legend and remove right hand side secondary y axis
   - Add a modal windows to configure each data line  been plotted or not and y axis upper / bottom limit ,also can switch       layout between grid and single layout.
2. Add `Euler` into main window menubar and remove `load euerl1d` from `app` menubar.

Plot:
截圖 2024-04-09 晚上11 05 52
截圖 2024-04-09 晚上10 24 58

### CODE REVIEW

File: modmesh/app/euler1d.py [hunk 1]
```diff
@@ -538,33 +693,13 @@ def single_layout(self):
...
-
-        options = QFileDialog.Options()
-        options |= QFileDialog.DontUseNativeDialog
-        fileName, _ = QFileDialog.getSaveFileName(None, "Save file", "",
-                                                  "All Files (*)",
-                                                  options=options)
-
-        if fileName != "":
-            fig.save(fileName, "JPG", 100)
-
```
Comment (j8xixo12) [inline modmesh/app/euler1d.py:567]: Using matplotlib built-in toolbar, it also contain save file button
so remove self-development save file callback.

File: modmesh/app/euler1d.py [hunk 2]
```diff
@@ -692,13 +842,27 @@ def content(self):
...
+                  .close(self.on_close)):
+                with VBox():
+                    with HBox():
+                        Label("Layout selection")
+                        Button("Grid").click(lambda: self.app.grid_layout())
+                        Button("Single").click((lambda:
+                                                self.app.single_layout()))
+                    Label("Data line configuration")
+                    Table(self.plot_config)
+                    Button("Save").click(lambda: self.app.update_layout())
```
Comment (j8xixo12) [inline modmesh/app/euler1d.py:876]: Add a modal window that allows the user to select which data line is plotted and adjust the upper/bottom limit of the y-axis. Additionally, it should be able to switch the plot layout.

File: modmesh/app/euler1d.py [hunk 3]
```diff
@@ -426,43 +583,22 @@ def build_single_figure(self):
...
+            if self.plot_config.get_var(data.name):
+                data.axis = ax
+                data.ana, = ax.plot(x, np.zeros_like(x),
+                                    f'{color}-',
+                                    label=f'{data.name}_ana')
+                data.num, = ax.plot(x, np.zeros_like(x),
+                                    f'{color}x',
+                                    label=f'{data.name}_num')
+                y_upper_lim_max = max(y_upper_lim_max, data.y_upper_lim)
+                y_bottom_lim_min = min(y_bottom_lim_min, data.y_bottom_lim)
```
Comment (j8xixo12) [inline modmesh/app/euler1d.py:604]: ditto.

File: modmesh/app/euler1d.py [hunk 4]
```diff
@@ -407,14 +570,8 @@ def build_single_figure(self):
...
-        ax.autoscale(enable=True, axis='y', tight=False)
-
-        # Matplotlib need to plot y axis on the left hand side first
-        # then the reset of axis can be plotted on right hand side
-        main_axis_plotted = False
-
-        # Record how many lines had been selected to plot
-        select_num = 0
+        y_upper_lim_max = 0.0
+        y_bottom_lim_min = sys.float_info.max
```
Comment (j8xixo12) [inline modmesh/app/euler1d.py:583]: Using the max and min limit as the plotting figure.

File: modmesh/app/euler1d.py [hunk 5]
```diff
@@ -78,6 +81,8 @@ class QuantityLine:
     ana: matplotlib.lines.Line2D = None
     num: matplotlib.lines.Line2D = None
     axis: matplotlib.pyplot.axis = None
+    y_upper_lim: float = 0.0
+    y_bottom_lim: float = 0.0
```
Comment (j8xixo12) [inline modmesh/app/euler1d.py:85]: added 2 float as Quantity Line's attribute to record current data line y axis upper / bottom limit.

File: cpp/modmesh/view/RManager.cpp [hunk 6]
```diff
@@ -131,6 +131,16 @@ void RManager::setUpMenu()
...
+        m_eulerMenu = m_mainWindow->menuBar()->addMenu(QString("Euler"));
+
+        {
+            auto app_name = QString("euler1d");
+            auto * action = new RAppAction(
+                QString("Load ") + app_name,
+                QString("Load ") + app_name,
+                QString("modmesh.app.") + app_name);
+            m_eulerMenu->addAction(action);
+        }
```
Comment (j8xixo12) [inline cpp/modmesh/view/RManager.cpp:143]: Add `Euler` into modmesh main window toolbar.

File: cpp/modmesh/view/RManager.cpp [hunk 7]
```diff
@@ -214,7 +224,6 @@ void RManager::setUpMenu()
         m_appMenu = m_mainWindow->menuBar()->addMenu(QString("App"));
 
         this->addApplication(QString("sample_mesh"));
-        this->addApplication(QString("euler1d"));
```
Comment (j8xixo12) [inline cpp/modmesh/view/RManager.cpp:217]: Since we have `Euler` in toolbar therefore removing euler1d from `App`.

**Review by j8xixo12 — COMMENTED** [review]
@yungyuc this patch is ready to review.

File: cpp/modmesh/view/RManager.cpp (hunk 6 above)
Comment (yungyuc) [reply cpp/modmesh/view/RManager.cpp:143]: This is good

File: cpp/modmesh/view/RManager.cpp (hunk 7 above)
Comment (yungyuc) [reply cpp/modmesh/view/RManager.cpp:217]: sweet

**Review by yungyuc — APPROVED** [review]
LGTM.  Since the CI mac runs are not stable, I will merged once it is healthy again

**Pull request merged**